source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+setup, but the entity map may also be modified during test execution via the
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ directive.
+
+Test runners MAY choose to implement the entity map in a fashion most suited to
+their language, but implementations MUST enforce both uniqueness of entity names
+and referential integrity when fetching an entity. Test runners MUST raise an
+error if an attempt is made to store an entity with a name that already exists
+in the map and MUST raise an error if an entity is not found for a name or is
+found but has an unexpected type.
+
+Consider the following examples::
+
+    # Error due to a duplicate name (client0 was already defined)
+    createEntities:
+      - client: { id: client0 }
+      - client: { id: client0 }
+
+    # Error due to a missing entity (client1 is undefined)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client1 }
+
+    # Error due to an unexpected entity type (session instead of client)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client0 }
+      - session: { id: session1, client: session0 }
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as a document by the test runner. This section defines the top-level
+keys for that document and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+- ``description``: Required string. The name of the test file.
+
+  This SHOULD describe the common purpose of tests in this file and MAY refer to
+  the filename (e.g. "updateOne-hint").
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies. Test runners will use this to determine compatibility
+  (i.e. whether and how the test file will be interpreted). The format of this
+  string is defined in `Version String`_; however, test files SHOULD NOT need to
+  refer to specific patch versions since patch-level changes SHOULD NOT alter
+  the structure of the test format (as previously noted in `Schema Version`_).
+
+.. _runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. If no requirements
+  are met, the test runner MUST skip this test file.
+
+  If set, the array should contain at least one document. The structure of each
```

Comment:
If we need it to be an array,  should we also open this up to take a single object?

Currently:

```js
{
  runOn: [
    {
      minServerVersion: "4.4",
    }
  ]
}
```

Optionally:

```js
{
  runOn: {
      minServerVersion: "4.4",
    }
}
```

Reply:
@prashantmital @isabelatkinson: Do you concur with always requiring arrays where we currently do "$type or array of $types"?

Reply:
Yeah in Rust this would require something along the lines of creating an enum that represents both formats (i.e. single `runOnRequirement` or array of `runOnRequirement`s) and then pattern matching over that to deal with both cases -- an unnecessary amount of complexity IMO. In general, I'd favor eliminating the "$type or array of $types" pattern to simplify the implementation of the runners (and minimize possible errors).

Reply:
I've replaced "$type or array of $types" with "array of $types" for the three places we did this: ``topologies``, ``observeEvents``, and ``ignoreCommandMonitoringEvents``.

Reply:
@jmikola I like this change, always having arrays be arrays makes things easier.